### ISSUE TITLE
when pushPage is fired in rapid succession

### ISSUE THREAD

**Issue by woyehenni** [OnsenUI/OnsenUI#349]
When i click one button(in page1) to push a new page(page2), just after the next page(page2) is indicated, click the button's position(Even though page1 is already hidden) once again in rapid succession on the second page(page2).
the second page(page2) would be push to page stack two times.
May be this is onsenui's bug?

Sorry for my poor english.

**Comment by argelius** [comment]
I don't want to change this behavior since I think it is correct. When `pushPage()` is called it should push a page regardless of what state it's in.

However, I'll add methods to detect state so you can do something like this:

```html
<ons-button 
  ng-attr-disabled="{{ navi.isPushing() }}" 
  ng-click="navi.pushPage('page.html')">
  Push
</ons-button>
```

I'll also add `ngDisabled` later so that can be used instead of `ngAttrDisabled`.

**Comment by Hirbod** [comment]
👍👍

**Comment by woyehenni** [comment]
Thanks.

**Comment by Hirbod** [comment]
Throwing a error on that fact is a bit hard, not? Same with page stack...

**Comment by Hirbod** [comment]
@argelius 

This new setting is horrible. I have a factory which will check for some settings and do a pushPage / resetToPage when the controller inits. Now I'll receive an error that throws an "Can't push when push is not done"... This is really UGLY and I don't want to use a timeout cause I want to prevent that the page beeing visible before. Please revert this!

**Comment by argelius** [comment]
Thanks @Hirbod ,

I will consider reverting this commit. 

The previous behavior before was probably safer. When calling `pushPage()` I guess it's best to push to the stack regardless of what state it's in so the behavior is predictable. 

The problem with pushing the same page twice may be annoying but is easily fixed and doesn't really break anything.

**Comment by Hirbod** [comment]
Hi @argelius,
Yes, this is exactly what I think. When someone need this, it's done with 2-3 lines of code (set a var, check postpush/prepush)

Maybe you should consider to add a setting var for this (for the navigator, something link blockpush="true") or something.

**Comment by argelius** [comment]
I reverted this. The old behavior seems safer and as @Hirbod said it can be solved pretty easily with some higher level code.

**Comment by Hirbod** [comment]
@argelius don't you think you could create a directive for this? Then users could add it to the navigator (just for comfortable reason)

**Comment by argelius** [comment]
Yes, we'll consider that for the next version. I think we'll keep the previous behavior for 1.2.2 since it's due soon.

The problem is that it introduces state to the navigator so we need to be a bit careful how we implement this the next time around.

**Comment by dojoVader** [comment]
Is there a way I can extend the NavigatorView and use it without modifying the core code. I'm experiencing this pushPage rendering the same page twice.

**Comment by argelius** [comment]
Hi @dojoVader !

Yes, it's an Angular service so you can change it if you like.

However, if you've discovered a bug please report it and we will try to fix it. If you've solved it yourself feel free to open a pull request.

**Comment by dojoVader** [comment]
This code managed to work for me once a door is locked i return a false by the time it executes subsequent call returns a false. But i used a decorator to monkey patch the NavigatorView.

function($provide, NavigatorView) {
        //Monkey patch all existing Angular Method. 
        //For the record this is the most Piece of shit concept of Monkey Patching why not override class
        //and use like Dojotoolkit
        $provide.decorator("NavigatorView", function($delegate) {
            $delegate.prototype.pushPage=function(page, options) {
                	if(this._doorLock.isLocked()){
                		return false;
                	}
                    if (this._profiling) {
                        console.time('pushPage');
                    }
                    options = options || {};
                    if (options && typeof options != 'object') {
                        throw new Error('options must be an object. You supplied ' + options);
                    }
                    if (this._emitPrePushEvent()) {
                        return;
                    }
                    this._doorLock.waitUnlock(function() {
                        this._pushPage(page, options);
                    }.bind(this));
                }
           
            return $delegate;
        })
    }

@woyehenni solution worked for me

**Comment by argelius** [comment]
@dojoVader 
Thanks for the code. However, I think if `pushPage()` is called twice it should push two pages.

What browser, device, OS, etc. are you using? Does a button click result in two click events?
What version of Onsen UI are you using? In the latest version we put a blocking mask on top of the navigator while animating to prevent it from being clicked twice.

**Comment by dojoVader** [comment]
Thanks for the reply I am using Onsen 1.2.2 

Test phone 1
Samsung Galaxy S3
Cordova Platform ( So am assuming it's the default Browser)
JellyBean

test Phone 2
Samsung Core 2
KitKat 
Cordova Platform

Once you click the button in a rapid succession, it loads the same page twice, since it is the same Navigator it is being called on,yes i agree that PushPage should be called twice, however the doorLock is unlocked once the PushPage has been resolved. Testing it now everything works.

However I do like the concept of masking as it's quite simple and totally eliminates any unforeseeable problem this code could cause.  

Thanks for the response

**Comment by argelius** [comment]
We released the  1.3.0-beta today, so if you want to try it you can install it with

```bash
$ bower install onsenui#1.3.0-beta
```

It contains the fix for navigator.

**Comment by rogervila** [comment]
Hello @argelius, I have the same problem as @dojoVader but with Onsen 1.3.16.
Should it be already solved?
